### PR TITLE
begin: add support for fetching remote ACIs to start from

### PR DESCRIPTION
**Pull request by dgonyeo**
With this change if the ACI provided to begin does not exist on the
filesystem, meta discovery is performed to locate and download the ACI.

### CODE REVIEW

**Comment by dgonyeo** [comment]
Addresses https://github.com/appc/acbuild/issues/17

File: acbuild/acbuild.go [hunk 1]
```diff
@@ -143,8 +143,26 @@ func runWrapper(cf func(cmd *cobra.Command, args []string) (exit int)) func(cmd
 		}
 		defer os.Remove(buildpath)
 
+		finfo, err := os.Stat(aciToModify)
+		if err != nil {
```
Comment (jonboulle) [inline acbuild/acbuild.go]: how about
```
switch {
case os.IsNotExist(err):
case err != nil:
case finfo.IsDir():
}
```

File: lib/begin.go [hunk 2]
```diff
@@ -112,3 +112,70 @@ func (a *ACBuild) Begin(start string) (err error) {
 
 	return nil
 }
+
+func (a *ACBuild) beginWithStart(start string, insecure bool) error {
```
Comment (jonboulle) [inline lib/begin.go]: `beginWithStart`?! has to be a better name. beginFromImage perhaps?

File: registry/fetch.go [hunk 3]
```diff
@@ -50,16 +50,25 @@ func (r Registry) tmpuncompressedpath() string {
 	return path.Join(r.DepStoreTarPath, "tmp.uncompressed.aci")
 }
 
-// FetchAndRender will fetch the given image and all of its dependencies if
-// they have not been fetched yet, and will then render them on to the
-// filesystem if they have not been rendered yet.
-func (r Registry) FetchAndRender(imagename types.ACIdentifier, labels types.Labels, size uint) error {
+func (r Registry) Fetch(imagename types.ACIdentifier, labels types.Labels, size uint, fetchDeps bool) error {
```
Comment (jonboulle) [inline registry/fetch.go]: // Fetch will foo a bar

File: registry/fetch.go [hunk 4]
```diff
@@ -50,16 +50,25 @@ func (r Registry) tmpuncompressedpath() string {
...
+	return err
+}
+
+// FetchAndRender will fetch the given image and all of its dependencies if
+// they have not been fetched yet, and will then render them on to the
+// filesystem if they have not been rendered yet.
+func (r Registry) FetchAndRender(imagename types.ACIdentifier, labels types.Labels, size uint) error {
+	err := r.Fetch(imagename, labels, size, true)
+	if err != nil {
+		return nil
```
Comment (jonboulle) [inline registry/fetch.go]: err?

File: lib/begin.go [hunk 5]
```diff
@@ -112,3 +112,70 @@ func (a *ACBuild) Begin(start string) (err error) {
...
+		return err
+	}
+
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return err
+	}
+
+	for _, file := range files {
+		if strings.HasPrefix(file.Name(), "sha512-") {
```
Comment (jonboulle) [inline lib/begin.go]: This seems very leaky with the registry, surely we can do better.

File: lib/begin.go [hunk 6]
```diff
@@ -112,3 +112,70 @@ func (a *ACBuild) Begin(start string) (err error) {
...
+			return util.UnTar(path.Join(dir, file.Name()), a.CurrentACIPath, nil)
+		}
+	}
+
+	return fmt.Errorf("aci missing after download?")
```
Comment (jonboulle) [inline lib/begin.go]: Probably panic here..?

**Comment by dgonyeo** [comment]
Should be ready for another pass.

File: lib/begin.go [hunk 7]
```diff
@@ -112,3 +112,70 @@ func (a *ACBuild) Begin(start string) (err error) {
...
+
+	files, err := ioutil.ReadDir(tmpDepStoreTarPath)
+	if err != nil {
+		return err
+	}
+
+	if len(files) == 0 {
+		panic("ACI is missing from temporary dependency store after download")
+	} else if len(files) > 1 {
+		panic("too many files in temporary dependency store after download")
```
Comment (jonboulle) [inline lib/begin.go]: print the files? :-)

Comment (jonboulle) [inline lib/begin.go]: how about just != 0

**Comment by jonboulle** [comment]
Assuming it works, code lgtm :-)

File: acbuild/acbuild.go [hunk 8]
```diff
@@ -143,8 +143,25 @@ func runWrapper(cf func(cmd *cobra.Command, args []string) (exit int)) func(cmd
...
+			return
+		case finfo.IsDir():
+			stderr("ACI to modify is a directory: %s.", aciToModify)
+			cmdExitCode = 1
+			return
+		}
+
 		a := lib.NewACBuild(buildpath, debug)
-		err = a.Begin(aciToModify)
+
```
Comment (jonboulle) [inline acbuild/acbuild.go]: sometimes I feel like you randomly insert empty lines just to troll or see if I notice..

Comment (dgonyeo) [inline acbuild/acbuild.go]: often after looking at something I get this feeling like it'd be easier to read if there was more whitespace, and so I put it in :p

**Pull request merged**